Sources/NIOHPACK/HPACKHeader.swift

```diff
@@ -276,42 +265,149 @@ public struct HPACKHeaders: ExpressibleByDictionaryLiteral, Sendable {
     /// - Returns: A list of the values for that header field name.
     @inlinable
     public subscript(canonicalForm name: String) -> [String] {
-        let result = self[name]
-        guard result.count > 0 else {
+        return self._values(forHeader: name, canonicalForm: true)
+    }
+
+    /// Return a sequence of header values for the header field named `name`.
+    ///
+    /// See also ``subscript(name:)`` and ``subscript(canonicalForm:)``.
+    ///
+    /// - Parameters:
+    ///   - name: The name of the header field whose values should be iterated.
+    ///   - canonicalForm: Whether the values should be decomposed into their "canonical form". That
+    ///     is, splitting them on commas as extensively as possible such that multiple values
+    ///     received on the one line are returned as separate entries.
+    /// - Note: Enabling `canonicalForm`  for the 'set-cookie' header has no efffect.
+    /// - Returns: A sequence of header values for `name`.
+    public func values(forHeader name: String, canonicalForm: Bool = false) -> Values {
+        if canonicalForm, name.isEqualCaseInsensitiveASCIIBytes(to: "set-cookie") {
+            // Not safe to split 'set-cookie'.
+            return Values(headers: self, name: name, canonicalize: false)
+        } else {
+            return Values(headers: self, name: name, canonicalize: canonicalForm)
+        }
+    }
+
+    /// Special internal function for use by tests.
+    internal subscript(position: Int) -> (String, String) {
+        precondition(position < self.headers.endIndex, "Position \(position) is beyond bounds of \(self.headers.endIndex)")
+        let header = self.headers[position]
+        return (header.name, header.value)
+    }
+
+    @inlinable
+    internal func _values(forHeader name: String, canonicalForm: Bool) -> [String] {
+        let values = self.values(forHeader: name, canonicalForm: canonicalForm)
+        var iterator = values.makeIterator()
+
+        guard let first = iterator.next() else {
+            // No value, no allocation.
             return []
         }
 
-        // It's not safe to split Set-Cookie on comma.
-        if name.isEqualCaseInsensitiveASCIIBytes(to: "set-cookie") {
-            return result
+        var headerValues = [String]()
+        // Avoid intermediate allocations by overcomitting on capacity. Note: may be an undercommit
+        // if the canonical form is being fetched.
+        headerValues.reserveCapacity(self.headers.count)
```

Comment:
Previously when getting headers with the `canonicalForm` we overcommitted by 4x the number of headers for the given name. Overcommitting by some multiple of `self.headers.count` feels like excessive resource usage. 

However this does mean we pay for intermediate reallocations if `canonicalForm: true` and the headers are dominated by the name we're looking for (which is why the alloc counting tests show no improvements.) I think this is totally fine though.

Reply:
Yeah I mean the only way we can fix this is to do the computation pre-pass, but having to scan the headers _twice_ to find the field names is extremely painful. The performance of that scan is really not good.